apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/db/derived/EntityDbTable.java

```diff
@@ -388,6 +396,42 @@ public void insert(T t) {
                     pstmt.executeUpdate();
                 }
             }
+            if (supportDelete()) {
+                // TODO replace 'height' selection from db by entity height when refactoring will be done
+                try (PreparedStatement maxHeightStatement = con.prepareStatement("SELECT MAX(height) from block");
+                    PreparedStatement thisExistsAndDeleted = con.prepareStatement("SELECT 1 from " + table + keyFactory.getPKClause() + " AND height = ? AND deleted = true")) {
+
+                    try (ResultSet maxHeightRs = maxHeightStatement.executeQuery()) {
+                        if (maxHeightRs.next()) {
+
+                            int index = dbKey.setPK(thisExistsAndDeleted, 1);
+                            int maxHeight = maxHeightRs.getInt(1);
+                            thisExistsAndDeleted.setInt(index, maxHeight);
+                            try (ResultSet rs = thisExistsAndDeleted.executeQuery()) {
+                                if (rs.next()) {
+                                    try (PreparedStatement selectPrevDeleted = con.prepareStatement("SELECT db_id FROM " + table + keyFactory.getPKClause() + " AND height < ? ORDER BY db_id DESC LIMIT 1");
+                                         PreparedStatement updatePrevDeleted = con.prepareStatement("UPDATE " + table + " SET deleted = false WHERE db_id = ?")) {
+                                        int selectPrevIndex = dbKey.setPK(selectPrevDeleted, 1);
+                                        selectPrevDeleted.setInt(selectPrevIndex, maxHeight);
+                                        try (ResultSet prevDeletedRs = selectPrevDeleted.executeQuery()) {
+                                            if (prevDeletedRs.next()) {
+                                                long prevDbId = prevDeletedRs.getLong(1);
+                                                updatePrevDeleted.setLong(1, prevDbId);
+                                                updatePrevDeleted.executeUpdate();
+                                            } else {
+                                                throw new IllegalStateException("Unable to find previous record for entity " + t + ", inconsistent database state (maybe 'delete' flow is broken)");
+                                            }
+                                        }
+                                    }
+                                }
+                            }
+                        } else {
```

Comment:
I suppose it would be good to dump 'context sensitive' information into log before raising exception.
Something like : **entity/table name, current block height, entity PK value, etc...**

Reply:
This exception should not be thrown in normal cases, but when thrown - it's easy to understand the real reason. Context data  is not needed, since such error does not depend on it.